polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -1349,33 +1373,60 @@ std::optional<MemoryAccessMatrix> MemoryAccessAnalysis::buildAccessMatrix(
 
   // Construct the memory access matrix. The number of rows is equal to the
   // dimensionality of the sycl.id used by the accessor subscript operation.
-  // The number of columns is equal to the number of loops surrounding the
-  // memory access plus the number of threads used in the kernel.
   MemoryAccessMatrix accessMatrix(sycl::getDimensions(accSubIndex.getType()),
-                                  loopAndThreadVars.size());
+                                  numColumns);
+
+  // Fill in a the access matrix element at [row,col] with the multiplier in
+  // 'expr' corresponding to the variable 'factor'.
+  auto createMatrixEntry = [&](Value expr, Value factor, size_t row,
+                               size_t col) -> bool {
+    assert(row >= 0 && row < accessMatrix.getNumRows() && "row out of bound");
+    assert(col >= 0 && col < accessMatrix.getNumColumns() &&
+           "col out of bound");
+
+    if (!usesValue(expr.getDefiningOp(), factor)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2) << "Doesn't use " << factor << "\n");
+      auto zero = ConstantIntRanges::constant(APInt());
+      accessMatrix(row, col) = IntegerValueRange(zero);
+      return true;
+    }
 
+    ValueOr<Multiplier> valOrMul = getMultiplier(expr, factor, solver);
+    if (!valOrMul.is<Multiplier>()) {
+      LLVM_DEBUG(llvm::dbgs().indent(2) << "Failed to find multiplier\n");
+      return false;
+    }
+
+    accessMatrix(row, col) = valOrMul.get<Multiplier>();
+    return true;
+  };
+
+  // Fill in the access matrix.
   for (size_t row = 0; row < accessMatrix.getNumRows(); ++row) {
     Value val = underlyingVals[row];
     LLVM_DEBUG(llvm::dbgs().indent(2)
                << "Analyzing underlying value: " << val << "\n");
 
-    for (size_t col = 0; col < accessMatrix.getNumColumns(); ++col) {
-      if (!usesValue(val.getDefiningOp(), loopAndThreadVars[col])) {
-        LLVM_DEBUG(llvm::dbgs().indent(2)
-                   << "Doesn't use " << loopAndThreadVars[col] << "\n");
+    // The leftmost 'threadVars.size()' matrix columns correspond to the
+    // thread variables, create the corresponding matrix entries.
+    for (size_t col = 0; col < numColumns - loopIVs.size(); ++col) {
+      if (col > threadVars.size() - 1) {
```

Comment:
need to actually know the get_global_id dimension, as `get_global_id(1)` can be used while `get_global_id(0)` is not.

Reply:
fixed thanks!